File: apps/api/services/connectors/runner.py
"""Parallel connector invocation with timeout and limits."""

import asyncio

from services.connectors.contract import (
    ConnectorInput,
    ConnectorOutput,
    ConnectorLimits,
    SourceMetadata,
)
from services.connectors.registry import get_adapter
from domain.integrations import load_config


def _error_metadata(source_id: str) -> SourceMetadata:
    return SourceMetadata(source_id=source_id, type="unknown", title=source_id, link=None)


async def fetch_one(
    source_id: str,
    tenant_id: str,
    query_text: str,
    limits: ConnectorLimits,
) -> ConnectorOutput:
    """Load config, get adapter, call fetch with timeout. Return output or error placeholder."""
    adapter = get_adapter(source_id)
    if not adapter:
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=_error_metadata(source_id),
            error="Adapter not found",
        )
    config = load_config(tenant_id, source_id)
    if not config:
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=_error_metadata(source_id),
            error="Integration not configured or disabled",
        )
    input_ = ConnectorInput(query_text=query_text, config=config, limits=limits)
    try:
        return await asyncio.wait_for(
            adapter.fetch(input_),
            timeout=float(limits.timeout_seconds),
        )
    except asyncio.TimeoutError:
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=_error_metadata(source_id),
            error="Timeout",
        )
    except Exception as e:
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=_error_metadata(source_id),
            error=str(e)[:200],
        )


async def fetch_all(
    source_ids: list[str],
    tenant_id: str,
    query_text: str,
    limits: ConnectorLimits | None = None,
) -> list[ConnectorOutput]:
    """Call all sources in parallel; return outputs in same order as source_ids."""
    limits = limits or ConnectorLimits()
    tasks = [fetch_one(sid, tenant_id, query_text, limits) for sid in source_ids]
    return list(await asyncio.gather(*tasks))

File: apps/web/lib/server/connectors/runner.ts
/**
 * Parallel connector invocation with timeout and limits. Server-only.
 */

import {
  type ConnectorInput,
  type ConnectorOutput,
  DEFAULT_LIMITS,
  type ConnectorLimits,
} from "./contract";
import { getAdapter } from "./registry";
import { loadConfig } from "../domain/integrations";

function errorMetadata(sourceId: string): ConnectorOutput["source_metadata"] {
  return {
    source_id: sourceId,
    type: "unknown",
    title: sourceId,
    link: null,
  };
}

export async function fetchOne(
  sourceId: string,
  tenantId: string,
  userId: string,
  queryText: string,
  limits: ConnectorLimits = DEFAULT_LIMITS
): Promise<ConnectorOutput> {
  const adapter = getAdapter(sourceId);
  if (!adapter) {
    return {
      success: false,
      fragments: [],
      source_metadata: errorMetadata(sourceId),
      error: "Adapter not found",
    };
  }
  const config = await loadConfig(tenantId, userId, sourceId);
  if (!config) {
    return {
      success: false,
      fragments: [],
      source_metadata: errorMetadata(sourceId),
      error: "Integration not configured or disabled",
    };
  }
  const input: ConnectorInput = {
    query_text: queryText,
    config: config,
    limits,
  };
  try {
    const timeoutMs = limits.timeout_seconds * 1000;
    const out = await Promise.race([
      adapter.fetch(input),
      new Promise<ConnectorOutput>((_, reject) =>
        setTimeout(() => reject(new Error("Timeout")), timeoutMs)
      ),
    ]);
    return out;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return {
      success: false,
      fragments: [],
      source_metadata: errorMetadata(sourceId),
      error: msg.slice(0, 200),
    };
  }
}

export async function fetchAll(
  sourceIds: string[],
  tenantId: string,
  userId: string,
  queryText: string,
  limits: ConnectorLimits = DEFAULT_LIMITS
): Promise<ConnectorOutput[]> {
  const results = await Promise.all(
    sourceIds.map((id) => fetchOne(id, tenantId, userId, queryText, limits))
  );
  return results;
}
